File: SourceAFIS/Meta/ObjectTree.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Reflection;

namespace SourceAFIS.Meta
{
    public sealed class ObjectTree
    {
        sealed class Item
        {
            public object Reference;
            public string Path;
        }

        Dictionary<string, Item> ByPath = new Dictionary<string,Item>();
        Dictionary<object, Item> ByReference = new Dictionary<object,Item>();

        public ObjectTree()
        {
        }

        public ObjectTree(object root)
        {
            Scan(root);
        }

        public ObjectTree(object root, string path)
        {
            Scan(root, path);
        }

        public void Scan(object root, string path = "")
        {
            if (!ByReference.ContainsKey(root) && !ByPath.ContainsKey(path))
            {
                Item item = new Item();
                item.Reference = root;
                item.Path = path;
                ByPath.Add(path, item);
                ByReference.Add(root, item);

                foreach (FieldInfo fieldInfo in root.GetType().GetFields())
                {
                    if (fieldInfo.GetCustomAttributes(typeof(NestedAttribute), true).Length > 0)
                    {
                        Scan(fieldInfo.GetValue(root), path + (path != "" ? "." : "") + fieldInfo.Name);
                    }
                }
            }
        }

        public void Remove(object reference)
        {
            ByPath.Remove(GetPath(reference));
            ByReference.Remove(reference);
        }

        public void Remove(string path) { Remove(GetObject(path)); }

        public object GetObject(string path) { return ByPath[path].Reference; }
        public string GetPath(object reference) { return ByReference[reference].Path; }
        public bool Contains(string path) { return ByPath.ContainsKey(path); }
        public bool Contains(object reference) { return ByReference.ContainsKey(reference); }
        public object[] GetAllObjects() { return new List<object>(ByReference.Keys).ToArray(); }
        public string[] GetAllPaths() { return new List<string>(ByPath.Keys).ToArray(); }
    }
}

File: java/sourceafis/src/sourceafis/meta/ObjectTree.java
package sourceafis.meta;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.IdentityHashMap;

public class ObjectTree {
	private static class Item {
		public String path;
		public Object reference;
	}
	
	private HashMap<String, Item> byPath = new HashMap<String, Item>();
	private IdentityHashMap<Object, Item> byReference = new IdentityHashMap<Object, Item>();

	public ObjectTree() {}
	public ObjectTree(Object root) { scan(root); }
	public ObjectTree(Object root, String path) { scan(root, path); }
	public void scan(Object root) { scan(root, ""); }
	
	public void scan(Object root, String path) {
		if (!byReference.containsKey(root) && !byPath.containsKey(path)) {
			Item item = new Item();
			item.path = path;
			item.reference = root;
			byPath.put(path, item);
			byReference.put(root, item);
			
			try {
				for (Field field : root.getClass().getFields())
					if (field.isAnnotationPresent(Nested.class))
						scan(field.get(root), path + (!path.isEmpty() ? "." : "") + field.getName());
			} catch (IllegalAccessException e) {
				throw new RuntimeException(e);
			}
		}
	}
	
	public Object getObject(String path) { return byPath.get(path).reference; }
	public String getPath(Object reference) { return byReference.get(reference).path; }
	public ArrayList<Object> getAllObjects() { return new ArrayList<Object>(byReference.keySet()); }
	public ArrayList<String> getAllPaths() { return new ArrayList<String>(byPath.keySet()); }
	public boolean containsObject(Object reference) { return byReference.containsKey(reference); }
	public boolean containsPath(String path) { return byPath.containsKey(path); }
}
